core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,815 @@
+
+    languageCode = questionPlayerEngineService.getLanguageCode();
+    expect(languageCode).toBe('ab');
+  });
+
+  it('should always return false when calling \'isInPreviewMode()\'', () => {
+    let previewMode = questionPlayerEngineService.isInPreviewMode();
+
+    expect(previewMode).toBe(false);
+  });
+
+  it('should show warning message while loading a question ' +
+    'if the question name is empty', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    singleQuestionBackendDict.question_state_data
+      .content.html = null;
+    let alertsServiceSpy = spyOn(
+      alertsService, 'addWarning').and.returnValue();
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => html);
+
+    questionPlayerEngineService.init(
+      [singleQuestionBackendDict], successHandler, failHandler);
+
+    expect(alertsServiceSpy).toHaveBeenCalledWith(
+      'Question name should not be empty.');
+  });
+
+  it('should not load questions if there are no questions', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    let alertsServiceSpy = spyOn(
+      alertsService, 'addWarning').and.returnValue();
+
+    questionPlayerEngineService.init(
+      [], successHandler, failHandler);
+
+    expect(alertsServiceSpy).toHaveBeenCalledWith(
+      'There are no questions to display.');
+    expect(successHandler).not.toHaveBeenCalled();
+    expect(failHandler).toHaveBeenCalled();
+  });
+
+  describe('on submitting answer ', () => {
+    it('should call success callback if the submitted' +
+      'answer is correct', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+
+      questionPlayerEngineService.init(
+        multipleQuestionsBackendDict, successHandler, failHandler);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(questionPlayerEngineService.isAnswerBeingProcessed()).toBe(false);
+      expect(successCallback).toHaveBeenCalled();
+    });
+
+    it('should not submit answer again if the answer ' +
+      'is already being processed', () => {
+      let successCallback = jasmine.createSpy('success');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
```

Comment:
Although this is not really used because `answerClassificationService.getMatchingClassificationResult` gets mocked, can we still initialise this variable with a dummy value? It seems odd to pass around `undefined` in a variable.

Reply:
done

Reply:
Not done? `interactionRulesService` does not seem to be initialised.

Reply:
sorry updated now.